spec.md

```diff
@@ -244,6 +244,58 @@ A common use of an externally defined module is to represent a Verilog module
 that will be written separately and provided together with FIRRTL-generated
 Verilog to downstream tools.
 
+## Functions
+
+A function has a name, list of arguments, and a list of statements representing
+a data-flow graph of stateless circuit components.  Each argument is specified 
+by its direction, which may be input or output, a name, and the data type of the 
+argument.
+
+Functions participate in width-inference (see [@sec:width-inference]) at 
+invocation sites in the same way modules to at instantiation sites.
+
+Functions cannot contain memories, registers, or instances.  This makes them
+"stateless"
+
+Functions can optional specify that they are "pure" functions.  These do not
+have side-effecting operations (such as printf).
+
+
+``` firrtl
+function MyPureFunction pure :
+  input foo : UInt
+  output bar : UInt
+  bar <= foo
+
+function MyFunction :
+  input clk: Clock
+  input halt: UInt<1>
+  stop(clk, halt, 42) : optional_name
+```
+
+## Intrinsics
+
+An intrinsic is a function. Intrinsics have all the components of a function
+except  a list of statements.  An intrinsic is a firrtl implementation provided 
+function.  The name of intrinsics are implementation defined.  A firrtl 
+implementation shall reject IR with unknown intrinsics.  Being functions,
+intrinsics are stateless.
+
+The widths of all intrinsic ports must be specified.  Width inference, described 
+in [@sec:width-inference], is not supported.
+
+Intrinsics can be "pure".  These do not have side-effects.  Whether a particular 
+intrinsic is pure or not is implementation defined and implementations shall 
+check that intrinsics declarations have the expected arguments, types, and 
+purity.
+
+``` firrtl
+intrinsic synopsys.4mux.u8 pure :
+  input cond : UInt<2>
+  output bar : Vector<4, UInt<8>>
+```
```

Comment:
Are there use cases for intrinsics with module semantics? Something like memories, certain forms of registers, etc.. It seems like the intrinsic above could cover that, but I'm not entirely sure. E.g., what happens if someone puts a `reg` intside a `function`?

Reply:
Isn't an intrinsic with module semantics an `extmodule`

Reply:
Yeah I guess you could do the whole thing as `extmodule`/`extfunction`. That's what LLVM does. Intrinsics are just external function declarations with a specific name. We could do `extmodule circt.intrinsic.synopsys.4mux.u8` instead. The good thing about an actual intrinsic construct is that you convey to the compiler that this thing _has_ to resolve to something the compiler internally knows, and the intrinsic doesn't accidentally show up in the SV output because the compiler didn't realize you wanted an intrinsic.

Reply:
sure... i think i get the idea/value of intrinsic, but maybe you could have intrinsic as a modifier of `extfunction` or `extmodule` to be more parallel

Reply:
> and the intrinsic doesn't accidentally show up in the SV output

I think we *do* want the intrinsic in the SV output, don't we? Otherwise we really are just re-defining `extmodule`.

Reply:
I'll let @darthscsi answer authoritatively but I'd expect this to be like LLVM intrinsics: https://llvm.org/docs/ExtendingLLVM.html#intrinsic-function .

Which is to say:
* intrinsics do not participate in linking and cannot be resolved to definitions in other files.  At FIRRTL level they're only ever declarations.  May be convenient to reject re-using their names elsewhere, or when linking (?).
* intrinsics defined entirely by the compiler (possibly emitting references to definitions linked in from support library (compiler-rt/libgcc, etc.) but doubt that's where we're going AFAIK)

Reply:
> I think we do want the intrinsic in the SV output, don't we?

We want the compiler to expand the intrinsic and do whatever it wants with it, and then the result of that should show up. But the intrinsic wont show up in SV as a module/function to represent the intrinsic. For example, that `synopsys.4mux.u8` you'd expect to expand to the actual multiplexers and necessary vendor pragmas to ensure they get inferred as 4-way muxes.

Reply:
The reason to make a distinction between modules and functions are that functions are stateless, which provides significant analysis power when dealing with them.  They are guaranteed to be data-flow like and behave as operations, not modules (and as a bonus, don't have to have connect semantics).

There is a use for module-like intrinsics.  Arguably, memories are already module-like intrinsics. Adding generic support for this is outside the scope of this change, but you could imagine an intmodule parallel to intrinsic.

module = connect semantics, instance semantics, stateful
extmodule = module with missing body (and some other restrictions)
intmodule = compiler/implementation provided module (to be added when needed)
Parallels
function = dataflow semantics, op-like semantics, stateless (but may be side-effecting)
extfunction = function with missing body
intrinsic = function/implemenation provided function